crates/bevy_input_focus/src/directional_navigation.rs

```diff
@@ -0,0 +1,379 @@
+    /// as we must iterate over each entity to check for connections to the removed entity.
+    ///
+    /// If you are removing multiple entities, consider using [`remove_multiple`](Self::remove_multiple) instead.
+    pub fn remove(&mut self, entity: Entity) {
+        self.neighbors.remove(&entity);
+
+        for node in self.neighbors.values_mut() {
+            for neighbor in node.neighbors.iter_mut() {
+                if *neighbor == Some(entity) {
+                    *neighbor = None;
+                }
+            }
+        }
+    }
+
+    /// Removes a collection of entities from the navigation map.
+    ///
+    /// While this is still an O(n) operation, where n is the number of entities in the map,
+    /// it is more efficient than calling [`remove`](Self::remove) multiple times,
+    /// as we can check for connections to all removed entities in a single pass.
+    ///
+    /// An [`EntityHashSet`] must be provided as it is noticeably faster than the standard hasher or a [`Vec`].
+    pub fn remove_multiple(&mut self, entities: EntityHashSet) {
+        for entity in &entities {
+            self.neighbors.remove(entity);
+        }
+
+        for node in self.neighbors.values_mut() {
+            for neighbor in node.neighbors.iter_mut() {
+                if let Some(entity) = *neighbor {
+                    if entities.contains(&entity) {
+                        *neighbor = None;
+                    }
+                }
+            }
+        }
+    }
+
+    /// Completely clears the navigation map, removing all entities and connections.
+    pub fn clear(&mut self) {
+        self.neighbors.clear();
+    }
+
+    /// Adds an edge between two entities in the navigation map.
+    /// Any existing edge from A in the provided direction will be overwritten.
+    ///
+    /// The reverse edge will not be added, so navigation will only be possible in one direction.
+    /// If you want to add a symmetrical edge, use [`add_symmetrical_edge`](Self::add_symmetrical_edge) instead.
+    pub fn add_edge(&mut self, a: Entity, b: Entity, direction: CompassOctant) {
+        self.neighbors
+            .entry(a)
+            .or_insert(NavNeighbors::EMPTY)
+            .set(direction, b);
+    }
+
+    /// Adds a symmetrical edge between two entities in the navigation map.
+    /// The A -> B path will use the provided direction, while B -> A will use the [`CompassOctant::opposite`] variant.
+    ///
+    /// Any existing connections between the two entities will be overwritten.
+    pub fn add_symmetrical_edge(&mut self, a: Entity, b: Entity, direction: CompassOctant) {
+        self.add_edge(a, b, direction);
+        self.add_edge(b, a, direction.opposite());
+    }
+
+    /// Add symettrical edges between all entities in the provided slice, looping back to the first entity at the end.
+    ///
+    /// This is useful for creating a circular navigation path between a set of entities, such as a menu.
```

Comment:
Docs should note that these edges are symmetrical.

Reply:
They do? It's the second (misspelled) word.

Reply:
Never mind me, I just cannot read, my bad.